clientv3/balancer_test.go

```diff
@@ -122,3 +128,92 @@ func TestBalancerGetBlocking(t *testing.T) {
 		t.Errorf("Get() with no up endpoints should timeout, got %v", err)
 	}
 }
+
+// TestBalancerDoNotBlockOnClose ensures that balancer and grpc don't deadlock each other
+// due to rapid open/close conn. The deadlock causes balancer.Close() to block forever.
+// See issue: https://github.com/coreos/etcd/issues/7283 for more detail.
+func TestBalancerDoNotBlockOnClose(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	kcl := newKillConnListener(t, 2)
+	defer kcl.close()
+
+	for i := 0; i < 100; i++ {
+		sb := newSimpleBalancer(kcl.endpoints())
+		conn, err := grpc.Dial("", grpc.WithInsecure(),
+			grpc.WithBalancer(sb))
+		if err != nil {
+			t.Fatal(err)
+		}
+		<-sb.ConnectNotify()
+		closec := make(chan struct{})
+		// Close() might block on balancer.Close()
+		// if balancer and grpc deadlock each other.
+		go func() {
+			defer close(closec)
+			conn.Close()
+		}()
+		select {
+		case <-closec:
+		case <-time.After(time.Second):
+			t.Fatal("conn.Close() timeout")
+		}
+	}
+}
+
+// killConnListener listens incoming conn and kills it immediately.
+type killConnListener struct {
+	wg        sync.WaitGroup
+	eps       []string
+	stopc     chan struct{}
+	bindAddrc chan string
+	t         *testing.T
+}
+
+func newKillConnListener(t *testing.T, size int) *killConnListener {
+	kcl := &killConnListener{stopc: make(chan struct{}),
+		bindAddrc: make(chan string, size),
+		t:         t}
+
+	for i := 0; i < size; i++ {
+		ln, err := net.Listen("tcp", ":0")
+		if err != nil {
+			t.Fatal(err)
+		}
+		kcl.eps = append(kcl.eps, ln.Addr().String())
+		kcl.wg.Add(1)
+		go func(l net.Listener) {
```

Comment:
The `kcl.wg.Done()` won't work right-- it'll issue a `Done` before the goroutine spawned from kcl.listen(l) exits.

```go
go kcl.listen(ln)
```